### PR TITLE
Feature/expressions: added index() function

### LINKED ISSUE #1391: Expressions: add index() function
Add index() function that returns the index of the current entity (context node)

### PR DESCRIPTION
**Pull request by SteRiccio**
## Related Issue

Resolves #1391

## Brief description of the changes proposed an/or how the issue(s) has(have) been solved.

This PR introduces the custom function `index()` that returns the index of the specified node among the siblings of its parent entity.
It can be useful to define expressions using values from sibling entities. E.g. the default value of an attribute plot_id can be the value of the previous entity attriubute + 1: `cluster/plot[index(plot) -1].plot_id + 1`
Most of the changes are in recordExpressionParser and nodeDefExpressionValidator.
The evaluation of custom functions is done by passing an object with custom functions evaluators to the expression evaluator (evaluator.js).

## How has this been tested

Only unit tests have been defined, testing the function on a simple cluster/plot survey.

##### Do you consider this PR needs further testing?
- [x] No
- [ ] Yes

## Types of changes

- [ ] Docs change 
- [x] Refactoring 
- [ ] Dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Disclaimer

Further improvements on the expression evaluation/validation will be introduced in another pull request.

### CODE REVIEW

File: core/expressionParser/helpers/evaluator.js [hunk 1]
```diff
@@ -215,10 +193,10 @@ const typeFns = {
   [types.GroupExpression]: groupEval,
 }
 
-export const evalExpression = (expr, ctx) => {
-  const functions = R.pipe(R.prop('functions'), R.mergeRight(typeFns))(ctx)
+export const evalExpression = (expr, ctx = {}) => {
+  const expressionTypeEvaluators = R.pipe(R.prop('expressionTypeEvaluators'), R.mergeRight(typeFns))(ctx)
```
Comment (minotogna) [inline core/expressionParser/helpers/evaluator.js:197]: can we just call this evaluator ?

File: core/expressionParser/helpers/functionImplementations.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import * as A from '@core/arena'
+import * as DateUtils from '@core/dateUtils'
+
+import { functionNames } from './functions'
+
+export const functionImplementations = {
```
Comment (minotogna) [inline core/expressionParser/helpers/functionImplementations.js:6]: can we just add a property called ```impl``` to the object functions ?

File: core/expressionParser/helpers/functions.js [hunk 3]
```diff
@@ -1,10 +1,29 @@
-import * as A from '@core/arena'
-import * as DateUtils from '@core/dateUtils'
+export const functionNames = {
+  avg: 'avg',
+  count: 'count',
+  includes: 'includes',
+  index: 'index',
```
Comment (minotogna) [inline core/expressionParser/helpers/functions.js:5]: index should take as argument a node

File: test/unit/tests/002recordExpressionParser.test.js [hunk 4]
```diff
@@ -119,13 +116,18 @@ describe('RecordExpressionParser Test', () => {
     { q: 'plot[0].plot_multiple_number.length', r: 2 },
     { q: 'plot[1].plot_multiple_number.length', r: null },
     { q: 'plot[2].plot_multiple_number.length', r: 1 },
+    // entity index
+    { q: 'index()', r: 0, n: 'cluster/plot[0]/plot_id' },
+    { q: 'index()', r: 1, n: 'cluster/plot[1]/plot_id' },
+    { q: 'index()', r: 2, n: 'cluster/plot[2]/plot_id' },
```
Comment (minotogna) [inline test/unit/tests/002recordExpressionParser.test.js:122]: tests to pass should be:

```
    { q: 'index(plot)', r: 0, n: 'cluster/plot[0]' },
      { q: 'index(plot)', r: 1, n: 'cluster/plot[1]' },
      { q: 'index(cluster)', r: 0, n: 'cluster' },
      ```

File: webapp/components/expression/codemirrorArenaExpressionHint.js [hunk 5]
```diff
@@ -7,19 +7,20 @@ import * as ExpressionVariables from './expressionVariables'
...
   [Expression.modes.json]: {
-    isEmpty: `isEmpty(attribute_name) = true/false`,
-    min: 'max(3,1) = 1',
-    max: 'max(3,1,2) = 3',
-    pow: 'pow(2,3) = 2³ = 8',
-    ln: 'ln(10) = 2.302…',
-    log10: 'log10(100) = 2',
-    includes: `includes(multiple_attribute_name, 'value') = true/false`,
-    now: 'now()',
+    [Expression.functionNames.index]: `index()`,
```
Comment (minotogna) [inline webapp/components/expression/codemirrorArenaExpressionHint.js:10]: pass a variable_name as argument

**Review by minotogna — CHANGES_REQUESTED** [review]

File: core/expressionParser/helpers/evaluator.js (hunk 1 above)
Comment (SteRiccio) [reply core/expressionParser/helpers/evaluator.js:197]: ...evaluators of what?

File: core/expressionParser/helpers/functionImplementations.js (hunk 2 above)
Comment (SteRiccio) [reply core/expressionParser/helpers/functionImplementations.js:6]: we can, but if you want to pass a custom implementation, like I do from recordExpressionParser, you will have to pass the entire function descriptor.

File: core/expressionParser/helpers/functions.js (hunk 3 above)
Comment (SteRiccio) [reply core/expressionParser/helpers/functions.js:5]: ok, I'll change it

File: test/unit/tests/002recordExpressionParser.test.js (hunk 4 above)
Comment (SteRiccio) [reply test/unit/tests/002recordExpressionParser.test.js:122]: ok

File: test/utils/recordUtils.js [hunk 6]
```diff
@@ -31,9 +29,8 @@ export const getNodePath = node => (survey, record) => {
 
     if (NodeDef.isMultiple(nodeDef)) {
       const siblings = Record.getNodeChildrenByDefUuid(parentNode, nodeDefUuid)(record)
-      const index = R.findIndex(n => Node.getUuid(n) === Node.getUuid(node), siblings)
-      const position = index + 1
-      return `${parentNodePath}/${NodeDef.getName(nodeDef)}[${position}]`
+      const index = R.findIndex((n) => Node.getUuid(n) === Node.getUuid(node), siblings)
+      return `${parentNodePath}/${NodeDef.getName(nodeDef)}[${index}]`
```
Comment (ramirobg94) [inline test/utils/recordUtils.js:33]: before the text has:
[${position}] and position was index + 1
now the text id:
[${index}]

is still the same?

Comment (SteRiccio) [reply test/utils/recordUtils.js:33]: no...! I've changed it to be consistent with the way we access nodes now (zero based). Anyway, this module is used only for tests... probably other integration tests will be broken by these change, but they need in any case to be refactored.

File: webapp/components/expression/codemirrorArenaExpressionHint.js (hunk 5 above)
Comment (SteRiccio) [reply webapp/components/expression/codemirrorArenaExpressionHint.js:10]: right

File: core/expressionParser/helpers/evaluator.js (hunk 1 above)
Comment (SteRiccio) [reply core/expressionParser/helpers/evaluator.js:197]: ok for "evaluator"!

**Review by minotogna — APPROVED** [review]

**Pull request merged**